src/params_loader.rs

```diff
@@ -15,9 +17,82 @@ pub trait PipesParamsLoader {
     /// to create a PipesContext or should instead return a mock.
     fn is_dagster_pipes_process(&self) -> bool;
     /// Load params passed by the orchestration-side context injector.
-    fn load_context_params(&self) -> Map<String, Value>;
+    fn load_context_params(&self) -> Result<Map<String, Value>, ParamsError>;
     /// Load params passed by the orchestration-side message reader.
-    fn load_message_params(&self) -> Map<String, Value>;
+    fn load_message_params(&self) -> Result<Map<String, Value>, ParamsError>;
+}
+
+#[derive(Debug, Error)]
+#[error("{} parameter \"{}\" is {}", .origin, .param, .source)]
+pub struct ParamsError {
+    pub param: String,
+    pub origin: ParamOrigin,
+    pub source: ParamsErrorKind,
+}
+
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum ParamOrigin {
+    Cli,
+    EnvVar,
+}
+
+impl fmt::Display for ParamOrigin {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Cli => write!(f, "cli"),
+            Self::EnvVar => write!(f, "env var"),
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum ParamsErrorKind {
```

Comment:
This shares a subset of failures of `PayloadErrorKind`. 

I chose to kept them as distinct error types because:
- `PayloadErrorKind` can fail on IO
- The two error types may diverge over time